File: Simulation/Bronze/P3_2016-Mowing_the_Field/mowing.py
with open("mowing.in", 'r') as fin:
    N = int(fin.readline())
    instructions = [fin.readline().split() for _ in range(N)]

curr_cord = (0, 0)
visited = {curr_cord: 0}
time = 0
max_regrow = float('inf')

# Process each movement command
for direction, steps in instructions:
    # Determine movement delta based on direction
    dx = dy = 0
    if direction == 'N':
        dy = 1
    elif direction == 'S':
        dy = -1
    elif direction == 'E':
        dx = 1
    elif direction == 'W':
        dx = -1

    # Move in the specified direction for numSteps
    for _ in range(int(steps)):
        curr_cord = (curr_cord[0] + dx, curr_cord[1] + dy)
        time += 1

        # Check if the cell was visited before
        if curr_cord in visited:
            last_visit_time = visited[curr_cord]
            max_regrow = min(max_regrow, time - last_visit_time)

        # Update the last visit time for the current cell
        visited[curr_cord] = time

# Output the result, -1 if no revisit occurred
print(
    max_regrow if max_regrow != float('inf') else -1,
    file=open("mowing.out", 'w')
)

File: Simulation/Bronze/P3_2016-Mowing_the_Field/mowing.cpp
#include <bits/stdc++.h>

#define N_MAX 101

using namespace std;

int main() {
    freopen("mowing.in", "r", stdin);
    int n;
    cin >> n;
    pair<char, int> instructions[N_MAX];
    for (int i = 0; i < n; i++)
        cin >> instructions[i].first >> instructions[i].second;

    pair<int, int> currCord = {0, 0};
    map<pair<int, int>, int> visited;
    int time = 0;
    int maxRegrow = INT_MAX;

    // Process each movement command
    for (auto [direction, steps] : instructions) {
        // Determine movement delta based on direction
        int dx = 0, dy = 0;
        switch (direction) {
            case 'N':
                dy = 1;
                break;
            case 'S':
                dy = -1;
                break;
            case 'E':
                dx = 1;
                break;
            case 'W':
                dx = -1;
                break;
        }

        // Move in the specified direction for numSteps
        for (int si = 0; si < steps; si++) {
            currCord.first += dx;
            currCord.second += dy;
            time++;

            // Check if the cell was visited before
            if (visited.find(currCord) != visited.end()) {
                int lastVisitTime = visited[currCord];
                maxRegrow = min(maxRegrow, time - lastVisitTime);
            }

            // Update the last visit time for the current cell
            visited[currCord] = time;
        }
    }

    // Output the result, -1 if no revisit occurred
    freopen("mowing.out", "w", stdout);
    cout << (maxRegrow != INT_MAX ? maxRegrow : -1) << endl;
    return 0;
}
